src/backend/distributed/connection/connection_management.c

```diff
@@ -438,183 +466,372 @@ ShutdownConnection(MultiConnection *connection)
 
-			/*
-			 * Only sleep for a limited amount of time, so we can react to
-			 * interrupts in time, even if the platform doesn't interrupt
-			 * poll() after signal arrival.
-			 */
-			pollResult = poll(&pollFileDescriptor, 1, checkIntervalMS);
-#else /* !HAVE_POLL */
-			fd_set readFileDescriptorSet;
-			fd_set writeFileDescriptorSet;
-			fd_set exceptionFileDescriptorSet;
-			int selectTimeoutUS = checkIntervalMS * 1000;
-			struct timeval selectTimeout = { 0, selectTimeoutUS };
-			int selectFileDescriptor = PQsocket(connection->pgConn);
-
-			FD_ZERO(&readFileDescriptorSet);
-			FD_ZERO(&writeFileDescriptorSet);
-			FD_ZERO(&exceptionFileDescriptorSet);
-
-			if (pollmode == PGRES_POLLING_READING)
+			if (waitCount <= 0)
 			{
-				FD_SET(selectFileDescriptor, &readFileDescriptorSet);
+				break;
 			}
-			else if (pollmode == PGRES_POLLING_WRITING)
+		}
+
+		eventCount = WaitEventSetWait(waitEventSet, timeout, events, waitCount,
+									  WAIT_EVENT_CLIENT_READ);
+
+		for (eventIndex = 0; eventIndex < eventCount; eventIndex++)
+		{
+			WaitEvent *event = &events[eventIndex];
+			MultiConnectionState *connectionState =
+				(MultiConnectionState *) event->user_data;
+
+			if (event->events & WL_POSTMASTER_DEATH)
 			{
-				FD_SET(selectFileDescriptor, &writeFileDescriptorSet);
+				ereport(ERROR, (errmsg("postmaster was shut down, exiting")));
 			}
 
-			pollResult = (select) (selectFileDescriptor + 1, &readFileDescriptorSet,
-								   &writeFileDescriptorSet, &exceptionFileDescriptorSet,
-								   &selectTimeout);
-#endif /* HAVE_POLL */
-
-			if (pollResult == 0)
+			if (event->events & WL_LATCH_SET)
 			{
-				/*
-				 * Timeout exceeded. Two things to do:
-				 * - check whether any interrupts arrived and handle them
-				 * - check whether establishment for connection already has
-				 *   lasted for too long, stop waiting if so.
-				 */
+				ResetLatch(MyLatch);
+
 				CHECK_FOR_INTERRUPTS();
 
-				if (TimestampDifferenceExceeds(connection->connectionStart,
-											   GetCurrentTimestamp(),
-											   NodeConnectionTimeout))
+				if (InterruptHoldoffCount > 0 && (QueryCancelPending || ProcDiePending))
 				{
-					ereport(WARNING, (errmsg("could not establish connection after %u ms",
-											 NodeConnectionTimeout)));
-
-					/* close connection, otherwise we take up resource on the other side */
-					PQfinish(connection->pgConn);
-					connection->pgConn = NULL;
-					break;
+					/*
+					 * because we can't break from 2 loops easily we need to not forget to
+					 * reset the memory context
+					 */
+					MemoryContextDelete(MemoryContextSwitchTo(oldContext));
+					return;
 				}
+
+				continue;
 			}
-			else if (pollResult > 0)
-			{
-				/*
-				 * IO possible, continue connection establishment. We could
-				 * check for timeouts here as well, but if there's progress
-				 * there seems little point.
-				 */
-				break;
-			}
-			else
+
+			switch (MultiConnectionStatePoll(connectionState))
```

Comment:
The naming of `MultiConnectionStatePoll`  does not suggest it returns a value that is relevant to a wait event set. I'm also wondering whether this code could be simplified, e.g. 

```
stateChanged = MultiConnectionStatePoll(connectionState)
if (stateChanged)
{
    if (connectionState->phase == MULTI_CONNECTION_PHASE_CONNECTED ||
        connectionState->phase == MULTI_CONNECTION_PHASE_ERROR)
    {
        /* we cannot stop waiting for a connection, so rebuild the event set */
        waitEventSetRebuild = true;
    }
    else
    {
        /* connection state changed, reset the event mask */
        int eventMask = MultiConnectionStateEventMask(connectionState);
        ModifyWaitEvent(waitEventSet, event->pos, eventMask, NULL);
    }
}
```

Reply:
I like this approach a lot, implemented it with slight changes.